python-checks/src/main/java/org/sonar/python/checks/UseOfAnyAsTypeHintCheck.java

```diff
@@ -32,24 +45,81 @@
 public class UseOfAnyAsTypeHintCheck extends PythonSubscriptionCheck {
 
   private static final String MESSAGE = "Use a more specific type than `Any` for this type hint.";
+  private static final Set<String> OVERRIDE_FQNS = Set.of("typing.override", "typing.overload");
 
   @Override
   public void initialize(Context context) {
-    context.registerSyntaxNodeConsumer(Tree.Kind.RETURN_TYPE_ANNOTATION, UseOfAnyAsTypeHintCheck::checkForAnyInTypeHint);
-    context.registerSyntaxNodeConsumer(Tree.Kind.PARAMETER_TYPE_ANNOTATION, UseOfAnyAsTypeHintCheck::checkForAnyInTypeHint);
+    context.registerSyntaxNodeConsumer(Tree.Kind.RETURN_TYPE_ANNOTATION, UseOfAnyAsTypeHintCheck::checkForAnyInReturnTypeAndParameters);
+    context.registerSyntaxNodeConsumer(Tree.Kind.PARAMETER_TYPE_ANNOTATION, UseOfAnyAsTypeHintCheck::checkForAnyInReturnTypeAndParameters);
     context.registerSyntaxNodeConsumer(Tree.Kind.VARIABLE_TYPE_ANNOTATION, UseOfAnyAsTypeHintCheck::checkForAnyInTypeHint);
   }
 
   private static void checkForAnyInTypeHint(SubscriptionContext ctx) {
+    Optional.of((TypeAnnotation) ctx.syntaxNode())
+      .filter(UseOfAnyAsTypeHintCheck::isTypeAny)
+      .ifPresent(typeAnnotation -> ctx.addIssue(typeAnnotation.expression(), MESSAGE));
+  }
+
+  private static void checkForAnyInReturnTypeAndParameters(SubscriptionContext ctx) {
     TypeAnnotation typeAnnotation = (TypeAnnotation) ctx.syntaxNode();
-    if (isTypeAny(typeAnnotation)) {
-      ctx.addIssue(typeAnnotation.expression(), MESSAGE);
-    }
+    Optional.of(typeAnnotation)
+      .filter(UseOfAnyAsTypeHintCheck::isTypeAny)
+      .map(annotation -> (FunctionDef) TreeUtils.firstAncestorOfKind(annotation, Tree.Kind.FUNCDEF))
+      .filter(Predicate.not(UseOfAnyAsTypeHintCheck::hasFunctionOverrideOrOverloadDecorator))
+      .filter(Predicate.not(UseOfAnyAsTypeHintCheck::isParentFunctionAnOverrideWithoutDecorators))
+      .ifPresent(functionDef -> ctx.addIssue(typeAnnotation.expression(), MESSAGE));
   }
 
   private static boolean isTypeAny(@Nullable TypeAnnotation typeAnnotation) {
     return Optional.ofNullable(typeAnnotation)
       .map(annotation -> "typing.Any".equals(TreeUtils.fullyQualifiedNameFromExpression(annotation.expression())))
       .orElse(false);
   }
+
+  private static boolean hasFunctionOverrideOrOverloadDecorator(FunctionDef currentFunctionDef) {
+    return currentFunctionDef.decorators().stream()
+      .map(Decorator::expression)
+      .map(TreeUtils::fullyQualifiedNameFromExpression)
+      .anyMatch(OVERRIDE_FQNS::contains);
+  }
+
+  private static boolean isParentFunctionAnOverrideWithoutDecorators(FunctionDef currentMethodDef) {
```

Comment:
Could we achieve the same using the `FunctionSymbolImpl#owner` method? Though I know it's annoying as it's not part of the public API.

Also, maybe this is too lenient, but we have the helper `SymbolUtils#canBeAnOverridingMethod` which could also be used here. I assume we'd lose many issues, as it is much more lenient, but that might prevent quite a few FPs when overriding methods from other libraries.

Reply:
Yes this is great. Thank you!